pyiceberg/partitioning.py

```diff
@@ -215,3 +230,118 @@ def assign_fresh_partition_spec_ids(spec: PartitionSpec, old_schema: Schema, fre
+
+T = TypeVar("T")
+class PartitionSpecVisitor(Generic[T], ABC):
+    @abstractmethod
+    def identity(field_id: int, source_name: str, source_id: int) -> T:
+        """
+        Visit identity partition field
+        """
+
+    @abstractmethod
+    def bucket(field_id: int, source_name: str, source_id: int, num_buckets: int) -> T:
+        """
+        Visit bucket partition field
+        """
+
+    @abstractmethod
+    def truncate(field_id: int, source_name: str, source_id: int, width: int) -> T:
+        """
+        Visit truncate partition field
+        """
+
+    @abstractmethod
+    def year(field_id: int, source_name: str, source_id: int) -> T:
+        """
+        Visit year partition field
+        """
+
+    @abstractmethod
+    def month(field_id: int, source_name: str, source_id: int) -> T:
+        """
+        Visit month partition field
+        """
+
+    @abstractmethod
+    def day(field_id: int, source_name: str, source_id: int) -> T:
+        """
+        Visit day partition field
+        """
+
+    @abstractmethod
+    def hour(field_id: int, source_name: str, source_id: int) -> T:
+        """
+        Visit hour partition field
+        """
+
+    @abstractmethod
+    def always_null(field_id: int, source_name: str, source_id: int) -> T:
+        """
+        Visit void partition field
+        """
+
+    @abstractmethod
+    def unknown(field_id: int, source_name: str, source_id: int, transform: str) -> T:
+        """
+        Visit unknown partition field
+        """
+        raise ValueError(f"Unknown transform {transform} is not supported")
+
+class _PartitionNameGenerator(PartitionSpecVisitor[str]):
+    def identity(field_id: int, source_name: str, source_id: int) -> str:
+        return source_name
+    
+    def bucket(field_id: int, source_name: str, source_id: int, num_buckets: int) -> str:
+        return source_name + "_bucket_" + num_buckets
+
+    def truncate(field_id: int, source_name: str, source_id: int, width: int) -> str:
+        return source_name + "_trunc_" + width
+
+    def year(field_id: int, source_name: str, source_id: int) -> str:
+        return source_name + "_year"
+
+    def month(field_id: int, source_name: str, source_id: int) -> str:
+        return source_name + "_month"
+
+    def day(field_id: int, source_name: str, source_id: int) -> str:
+        return source_name + "_day"
+    
+    def hour(field_id: int, source_name: str, source_id: int) -> str:
+        return source_name + "_hour"
+
+    def always_null(field_id: int, source_name: str, source_id: int) -> str:
+        return source_name + "_null"
+
+R = TypeVar("R")
+
+@staticmethod
+def _visit(spec: PartitionSpec, schema: Schema, visitor: PartitionSpecVisitor[R]) -> [R]:
+    results = []
+    for field in spec.fields:
+        results.append(_visit(schema, field, visitor))
+    return results
+
+@staticmethod
+def _visit(schema: Schema, field: PartitionField, visitor: PartitionSpecVisitor[R]) -> [R]:
```

Comment:
It would be better to use `singledispatch` here. For example: 

https://github.com/apache/iceberg-python/blob/452238e33f2be86b9076be78db2441aea043afb8/pyiceberg/expressions/visitors.py#L139

This is faster (which surprised me at first, but it pushes it down to the C level) and looks nicer